How about replacement `jq` with `jaq`?

At the time of v0.1.0, the author deemed it most appropriate to use the original jq.

However, managing C-related stuff during build time had to be undertaken, and this was underestimated at the release of v0.1.0. After actually releasing and taking a look the installation error issues that were raised, it became apparent that continuing to use jq might not be the best decision.

Then, there is a project in Rust called [jaq](https://github.com/01mf02/jaq), which is a jq-clone. I hope to discuss whether or not to replace jq with jaq, and whether it is feasible to do so.

Here's a rough outline of the pros and cons:

[jq](https://github.com/jqlang/jq) ([j9](https://github.com/ynqa/j9))
- pros
  - As an original implementation, all filters defined here are, of course, available for use
- cons
  - Requires management of C dependencies, which might pose issues during build phase
    - https://github.com/ynqa/jnv/issues/6
    - https://github.com/ynqa/jnv/issues/13
    - https://github.com/ynqa/jnv/issues/1

[jaq](https://github.com/01mf02/jaq)
- pros
  - pure Rust, there is no need to manage C dependencies
- cons
  - jq-clone, not all filters may be supported
     - e.g. https://github.com/01mf02/jaq/issues/112
     - My concern is with this point. Ideally, it would be better if all features were available for use (Or maybe it's sufficient if only the basic filters are available...?)

I'm also having some issues compiling j9-sys

https://github.com/conda-forge/staged-recipes/pull/25806#issuecomment-2041434460

```
checking host system type... 
  --- stderr
  configure.ac:11: installing 'config/ar-lib'
  configure.ac:8: installing 'config/compile'
  configure.ac:19: installing 'config/config.guess'
  configure.ac:19: installing 'config/config.sub'
  configure.ac:9: installing 'config/install-sh'
  configure.ac:9: installing 'config/missing'
  Makefile.am: installing 'config/depcomp'
  configure.ac: installing 'config/ylwrap'
  parallel-tests: installing 'config/test-driver'
  Invalid configuration `/home/conda/staged-recipes/build_artifacts/jnv_1712488275842/_build_env/bin/x86_64-conda-linux-gnu': more than four components
  configure: error: /bin/sh /home/conda/staged-recipes/build_artifacts/jnv_1712488275842/work/target/x86_64-unknown-linux-gnu/release/build/j9-sys-b50a7ed31c9b8fd9/out/jq_build/config/config.sub /home/conda/staged-recipes/build_artifacts/jnv_1712488275842/_build_env/bin/x86_64-conda-linux-gnu failed
  thread 'main' panicked at /home/conda/staged-recipes/build_artifacts/jnv_1712488275842/_build_env/.cargo/registry/src/index.crates.io-6f17d22bba15001f/autotools-0.2.6/src/lib.rs:781:5:
```

Switching to `jaq` would fix this 😅

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -15,27 +15,6 @@ jobs:
             ~/.cargo/git
             target
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
-      - name: Install clang package
-        run: |
-          sudo apt-get update
-          sudo apt-get install -y \
-            build-essential \
-            autoconf \
-            libtool \
-            git \
-            wget \
-            software-properties-common
-          wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
-          sudo apt-add-repository "deb http://apt.llvm.org/bullseye/ llvm-toolchain-bullseye-15 main"
-          sudo apt-get update
-          sudo apt-get install -y \
-            clang-15 \
-            lldb-15 \
-            lld-15
-      - name: Set default compiler to clang
-        run: |
-          echo "CC=clang-15" >> $GITHUB_ENV
-          echo "CXX=clang++-15" >> $GITHUB_ENV
       - uses: actions-rs/toolchain@v1
         with:
           toolchain: stable

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -11,8 +11,10 @@ readme = "README.md"
 [dependencies]
 anyhow = "1.0.82"
 clap = { version = "4.5.4", features = ["derive"] }
-gag = "1.0.0"
-j9 = "0.1.3"
+jaq-core = "1.2.1"
+jaq-interpret = "1.2.1"
+jaq-parse = "1.0.2"
+jaq-std = "1.2.1"
 promkit = "0.4.3"
 radix_trie = "0.2.1"
 
@@ -32,11 +34,8 @@ installers = ["homebrew"]
 # A GitHub repo to push Homebrew formulas to
 tap = "ynqa/homebrew-tap"
 # Target platforms to build apps for (Rust target-triple syntax)
-targets = ["x86_64-apple-darwin", "x86_64-unknown-linux-gnu"]
+targets = ["aarch64-apple-darwin", "x86_64-apple-darwin", "x86_64-unknown-linux-gnu", "x86_64-pc-windows-msvc"]
 # Publish jobs to run in CI
 publish-jobs = ["homebrew"]
 # Publish jobs to run in CI
 pr-run-mode = "plan"
-
-[workspace.metadata.dist.dependencies.homebrew]
-automake = '*'

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "jnv"
-version = "0.2.3"
+version = "0.3.0"
 authors = ["ynqa <un.pensiero.vano@gmail.com>"]
 edition = "2021"
 description = "JSON navigator and interactive filter leveraging jq"

--- a/README.md
+++ b/README.md
@@ -14,6 +14,8 @@ and [jiq](https://github.com/fiatjaf/jiq).
 
 - Interactive JSON viewer and `jq` filter editor
   - Syntax highlighting for JSON
+  - Use [jaq]((https://github.com/01mf02/jaq)) parser for `jq` filters
+    - This eliminates the need for users to prepare `jq` on their own.
 - Capable of accommodating various format
   - Input: File, Stdin
   - Data: A JSON or multiple JSON structures
@@ -61,20 +63,9 @@ nix-shell -p jnv
 
 ### Cargo
 
-#### Requirements
-
-- [autoconf](https://www.gnu.org/software/autoconf/)
-- [automake](https://www.gnu.org/software/automake/)
-- [libtool](https://www.gnu.org/software/libtool/)
-- [clang](https://clang.llvm.org/)
-
 ```bash
 cargo install jnv
 ```
-
-> [!NOTE]
-> *jnv* does not require users to install `jq` on their system,
-> because it utilizes [j9](https://github.com/ynqa/j9) Rust bindings.
 
 ## Examples
 

--- a/README.md
+++ b/README.md
@@ -16,6 +16,20 @@ and [jiq](https://github.com/fiatjaf/jiq).
   - Syntax highlighting for JSON
   - Use [jaq](https://github.com/01mf02/jaq) to apply `jq` filter
     - This eliminates the need for users to prepare `jq` on their own.
+
+> [!IMPORTANT]
+> Starting from v0.3.0, the transition from libjq Rust binding
+> [j9](https://github.com/ynqa/j9) to jq clone
+> [jaq](https://github.com/01mf02/jaq) was made.
+>
+> This change eliminated the need to manage C-related dependencies
+> that include external tools like autoconf, thus simplifying the build process.
+> However, please note that some filters are not yet supported by jaq.
+> For more details, refer to GitHub issue
+> [#24](https://github.com/ynqa/jnv/issues/24).
+>
+> Please continue to provide feedback regarding this transition.
+
 - Capable of accommodating various format
   - Input: File, Stdin
   - Data: A JSON or multiple JSON structures

--- a/README.md
+++ b/README.md
@@ -14,7 +14,7 @@ and [jiq](https://github.com/fiatjaf/jiq).
 
 - Interactive JSON viewer and `jq` filter editor
   - Syntax highlighting for JSON
-  - Use [jaq]((https://github.com/01mf02/jaq)) parser for `jq` filters
+  - Use [jaq](https://github.com/01mf02/jaq) to apply `jq` filter
     - This eliminates the need for users to prepare `jq` on their own.
 - Capable of accommodating various format
   - Input: File, Stdin
@@ -118,39 +118,23 @@ Examples:
         cat data.json | jnv
 
 Arguments:
-  [INPUT]
-          Optional path to a JSON file. If not provided or if "-" is specified, reads from standard input
+  [INPUT]  Optional path to a JSON file. If not provided or if "-" is specified, reads from standard input
 
 Options:
   -e, --edit-mode <EDIT_MODE>
-                  Specifies the edit mode for the interface.
-                  Acceptable values are "insert" or "overwrite".
-                  - "insert" inserts a new input at the cursor's position.
-                  - "overwrite" mode replaces existing characters with new input at the cursor's position.
-          [default: insert]
-
+          Edit mode for the interface ('insert' or 'overwrite'). [default: insert]
   -i, --indent <INDENT>
-                  Affect the formatting of the displayed JSON,
-                  making it more readable by adjusting the indentation level.
-          [default: 2]
-
+          Number of spaces used for indentation in the visualized data. [default: 2]
   -n, --no-hint
-                  When this option is enabled, it prevents the display of
-                  hints that typically guide or offer suggestions to the user.
-
-  -d, --expand-depth <EXPAND_DEPTH>
-                  Specifies the initial depth to which JSON nodes are expanded in the visualization.
-                  Note: Increasing this depth can significantly slow down the display for large datasets.
-          [default: 3]
-
+          Disables the display of hints.
+  -d, --expand-depth <JSON_EXPAND_DEPTH>
+          Initial depth to which JSON nodes are expanded in the visualization. [default: 3]
+  -s, --limit-length <JSON_LIMIT_LENGTH>
+          Limit length of JSON array in the visualization. [default: 50]
   -l, --suggestion-list-length <SUGGESTION_LIST_LENGTH>
-                  Controls the number of suggestions displayed in the list,
-                  aiding users in making selections more efficiently.
-          [default: 3]
-
+          Number of suggestions visible in the list. [default: 3]
   -h, --help
-          Print help (see a summary with '-h')
-
+          Print help (see more with '--help')
   -V, --version
           Print version
 ```

--- a/src/jnv.rs
+++ b/src/jnv.rs
@@ -1,7 +1,8 @@
 use std::cell::RefCell;
 
 use anyhow::Result;
-use gag::Gag;
+
+use jaq_interpret::{Ctx, FilterT, ParseCtx, RcIter, Val};
 
 use promkit::{
     crossterm::{
@@ -11,7 +12,7 @@ use promkit::{
     json::{self, JsonNode, JsonPathSegment, JsonStream},
     listbox,
     pane::Pane,
-    serde_json::{self, Deserializer},
+    serde_json,
     snapshot::Snapshot,
     style::StyleBuilder,
     suggest::Suggest,
@@ -23,64 +24,37 @@ use crate::trie::FilterTrie;
 
 mod keymap;
 
-/// Deserializes a JSON string into a vector of `serde_json::Value`.
-///
-/// This function takes a JSON string as input and attempts to parse it into a vector
-/// of `serde_json::Value`, which represents any valid JSON value (e.g., object, array, string, number).
-/// It leverages `serde_json::Deserializer` to parse the string and collect the results.
-///
-/// # Arguments
-/// * `json_str` - A string slice that holds the JSON data to be deserialized.
-///
-/// # Returns
-/// An `anyhow::Result` wrapping a vector of `serde_json::Value`. On success, it contains the parsed
-/// JSON data. On failure, it contains an error detailing what went wrong during parsing.
-fn deserialize_json(
-    json_str: &str,
-    limit_length: Option<usize>,
+fn run_jaq(
+    query: &str,
+    json_stream: Vec<serde_json::Value>,
 ) -> anyhow::Result<Vec<serde_json::Value>> {
-    let deserializer = Deserializer::from_str(json_str).into_iter::<serde_json::Value>();
-    let results = match limit_length {
-        Some(l) => deserializer.take(l).collect::<Result<Vec<_>, _>>(),
-        None => deserializer.collect::<Result<Vec<_>, _>>(),
-    };
-    results.map_err(anyhow::Error::from)
-}
+    let mut ret = Vec::<serde_json::Value>::new();
 
-fn run_jq(query: &str, json_stream: &[serde_json::Value]) -> anyhow::Result<Vec<String>> {
-    // libjq writes to the console when an internal error occurs.
-    //
-    // e.g.
-    // ```
-    // let _ = j9::run(". | select(.number == invalid_no_quote)", "{}");
-    // jq: error: invalid_no_quote/0 is not defined at <top-level>, line 1:
-    //     . | select(.number == invalid_no_quote)
-    // ```
-    //
-    // While errors themselves are not an issue,
-    // they interfere with the console output handling mechanism
-    // in promkit and qjq (e.g., causing line numbers to shift).
-    // Therefore, we'll ignore console output produced inside j9::run.
-    //
-    // It's possible that this could be handled
-    // within github.com/ynqa/j9, but for now,
-    // we'll proceed with this workaround.
-    //
-    // For reference, the functionality of a quiet mode in libjq is
-    // also being discussed at https://github.com/jqlang/jq/issues/1225.
-    let ignore_err = Gag::stderr().unwrap();
-    let mut jq_ret = Vec::<String>::new();
-    for v in json_stream.iter() {
-        let inner_ret: Vec<String> = match j9::run(query, &v.to_string()) {
-            Ok(ret) => ret,
-            Err(e) => {
-                return Err(anyhow::anyhow!(e));
-            }
-        };
-        jq_ret.extend(inner_ret);
+    for input in json_stream {
+        let mut ctx = ParseCtx::new(Vec::new());
+        ctx.insert_natives(jaq_core::core());
+        ctx.insert_defs(jaq_std::std());
+
+        let (f, errs) = jaq_parse::parse(query, jaq_parse::main());
+        if !errs.is_empty() {
+            let error_message = errs
+                .iter()
+                .map(|e| e.to_string())
+                .collect::<Vec<_>>()
+                .join(", ");
+            return Err(anyhow::anyhow!(error_message));
+        }
+
+        let f = ctx.compile(f.unwrap());
+        let inputs = RcIter::new(core::iter::empty());
+        let mut out = f.run((Ctx::new([], &inputs), Val::from(input)));
+
+        while let Some(Ok(val)) = out.next() {
+            ret.push(val.into());
+        }
     }
-    drop(ignore_err);
-    Ok(jq_ret)
+
+    Ok(ret)
 }
 
 pub struct JsonTheme {
@@ -132,24 +106,20 @@ pub struct Jnv {
     suggest: Suggest,
 
     json_expand_depth: Option<usize>,
-    json_limit_length: Option<usize>,
     no_hint: bool,
 }
 
 impl Jnv {
     #[allow(clippy::too_many_arguments)]
     pub fn try_new(
-        input: String,
+        input_stream: Vec<serde_json::Value>,
         filter_editor: text_editor::State,
         hint_message: text::State,
         suggestions: listbox::State,
         json_theme: JsonTheme,
         json_expand_depth: Option<usize>,
-        json_limit_length: Option<usize>,
         no_hint: bool,
     ) -> Result<Prompt<Self>> {
-        let input_stream = deserialize_json(&input, json_limit_length)?;
-
         let mut trie = FilterTrie::default();
         trie.insert(".", input_stream.clone());
 
@@ -209,7 +179,6 @@ impl Jnv {
                 trie,
                 suggest,
                 json_expand_depth,
-                json_limit_length,
                 no_hint,
                 input_stream,
             },
@@ -284,7 +253,7 @@ impl promkit::Renderer for Jnv {
                     );
                 }
                 None => {
-                    match run_jq(&filter, &self.input_stream) {
+                    match run_jaq(&filter, self.input_stream.clone()) {
                         Ok(ret) => {
                             if ret.is_empty() {
                                 self.update_hint_message(
@@ -302,53 +271,30 @@ impl promkit::Renderer for Jnv {
                                         JsonStream::new(searched.clone(), self.json_expand_depth);
                                 }
                             } else {
-                                match deserialize_json(&ret.join("\n"), self.json_limit_length) {
-                                    Ok(jsonl) => {
-                                        let stream =
-                                            JsonStream::new(jsonl.clone(), self.json_expand_depth);
+                                let stream = JsonStream::new(ret.clone(), self.json_expand_depth);
 
-                                        let is_null = stream.roots().iter().all(|node| {
-                                            node == &JsonNode::Leaf(serde_json::Value::Null)
-                                        });
-                                        if is_null {
-                                            self.update_hint_message(
-                                                format!("JSON query resulted in 'null', which may indicate a typo or incorrect query: '{}'", &filter),
-                                                StyleBuilder::new()
-                                                    .fgc(Color::Yellow)
-                                                    .attrs(Attributes::from(Attribute::Bold))
-                                                    .build(),
-                                            );
-                                            if let Some(searched) = self.trie.prefix_search(&filter)
-                                            {
-                                                self.json.stream = JsonStream::new(
-                                                    searched.clone(),
-                                                    self.json_expand_depth,
-                                                );
-                                            }
-                                        } else {
-                                            // SUCCESS!
-                                            self.trie.insert(&filter, jsonl);
-                                            self.json.stream = stream;
-                                        }
-                                    }
-                                    Err(e) => {
-                                        self.update_hint_message(
-                                            format!(
-                                                "Failed to parse query result for viewing: {}",
-                                                e
-                                            ),
-                                            StyleBuilder::new()
-                                                .fgc(Color::Red)
-                                                .attrs(Attributes::from(Attribute::Bold))
-                                                .build(),
+                                let is_null = stream
+                                    .roots()
+                                    .iter()
+                                    .all(|node| node == &JsonNode::Leaf(serde_json::Value::Null));
+                                if is_null {
+                                    self.update_hint_message(
+                                        format!("JSON query resulted in 'null', which may indicate a typo or incorrect query: '{}'", &filter),
+                                        StyleBuilder::new()
+                                            .fgc(Color::Yellow)
+                                            .attrs(Attributes::from(Attribute::Bold))
+                                            .build(),
+                                    );
+                                    if let Some(searched) = self.trie.prefix_search(&filter) {
+                                        self.json.stream = JsonStream::new(
+                                            searched.clone(),
+                                            self.json_expand_depth,
                                         );
-                                        if let Some(searched) = self.trie.prefix_search(&filter) {
-                                            self.json.stream = JsonStream::new(
-                                                searched.clone(),
-                                                self.json_expand_depth,
-                                            );
-                                        }
                                     }
+                                } else {
+                                    // SUCCESS!
+                                    self.trie.insert(&filter, ret);
+                                    self.json.stream = stream;
                                 }
                             }
                         }

--- a/src/main.rs
+++ b/src/main.rs
@@ -11,6 +11,7 @@ use clap::Parser;
 use promkit::{
     crossterm::style::{Attribute, Attributes, Color},
     listbox,
+    serde_json::{self, Deserializer},
     style::StyleBuilder,
     text, text_editor,
 };
@@ -158,10 +159,35 @@ fn parse_input(args: &Args) -> Result<String> {
     Ok(ret)
 }
 
+/// Deserializes a JSON string into a vector of `serde_json::Value`.
+///
+/// This function takes a JSON string as input and attempts to parse it into a vector
+/// of `serde_json::Value`, which represents any valid JSON value (e.g., object, array, string, number).
+/// It leverages `serde_json::Deserializer` to parse the string and collect the results.
+///
+/// # Arguments
+/// * `json_str` - A string slice that holds the JSON data to be deserialized.
+///
+/// # Returns
+/// An `anyhow::Result` wrapping a vector of `serde_json::Value`. On success, it contains the parsed
+/// JSON data. On failure, it contains an error detailing what went wrong during parsing.
+fn deserialize_json(
+    json_str: &str,
+    limit_length: Option<usize>,
+) -> anyhow::Result<Vec<serde_json::Value>> {
+    let deserializer = Deserializer::from_str(json_str).into_iter::<serde_json::Value>();
+    let results = match limit_length {
+        Some(l) => deserializer.take(l).collect::<Result<Vec<_>, _>>(),
+        None => deserializer.collect::<Result<Vec<_>, _>>(),
+    };
+    results.map_err(anyhow::Error::from)
+}
+
 fn main() -> Result<()> {
     let args = Args::parse();
 
     let input = parse_input(&args)?;
+    let input_stream = deserialize_json(&input, args.json_limit_length)?;
 
     let filter_editor = text_editor::State {
         texteditor: Default::default(),
@@ -216,13 +242,12 @@ fn main() -> Result<()> {
     };
 
     let mut prompt = Jnv::try_new(
-        input,
+        input_stream,
         filter_editor,
         hint_message,
         suggestions,
         json_theme,
         args.json_expand_depth,
-        args.json_limit_length,
         args.no_hint,
     )?;
     let _ = prompt.run()?;

